ocaml/xapi/message_forwarding.ml

```diff
@@ -161,25 +165,37 @@ let log_exn_ignore ?(doc = "performing unknown operation") f x =
 
 
 (* Given an SR, return a PBD to use for some storage operation. *)
-(* In the case of SR.destroy we need to be able to forward the SR operation when all
-   PBDs are unplugged - this is the reason for the consider_unplugged_pbds optional
-   argument below. All other SR ops only consider plugged PBDs... *)
-let choose_pbd_for_sr ?(consider_unplugged_pbds=false) ~__context ~self () =
+(* In the case of SR.destroy and forget we need to be able to forward the SR
+   operation when all PBDs are unplugged. This is the reason for the
+   consider_unplugged_pbds optional argument below. All other SR ops only
+   consider plugged PBDs. *)
+let choose_pbds_for_sr ?(consider_unplugged_pbds=false) ~__context ~self () =
   let all_pbds = Db.SR.get_PBDs ~__context ~self in
-  let plugged_pbds = List.filter (fun pbd->Db.PBD.get_currently_attached ~__context ~self:pbd) all_pbds in
+  let plugged_pbds = List.filter (fun self -> Db.PBD.get_currently_attached ~__context ~self) all_pbds in
   let pbds_to_consider = if consider_unplugged_pbds then all_pbds else plugged_pbds in
-  if Helpers.is_sr_shared ~__context ~self then
-    let master = Helpers.get_master ~__context in
-    let master_pbds = Db.Host.get_PBDs ~__context ~self:master in
-    (* shared SR operations must happen on the master *)
-    match Listext.intersect pbds_to_consider master_pbds with
-    | pbd :: _ -> pbd (* ok, master plugged *)
-    | [] -> raise (Api_errors.Server_error(Api_errors.sr_no_pbds, [ Ref.string_of self ])) (* can't do op, master pbd not plugged *)
-  else
-    match pbds_to_consider with
-    | [] -> raise (Api_errors.Server_error(Api_errors.sr_no_pbds, [ Ref.string_of self ]))
-    | pdb :: _ -> pdb
-
+  let pbds = PBDSet.of_list pbds_to_consider in
+  let pbd_candidates =
+    if Helpers.is_sr_shared ~__context ~self then
+      let master = Helpers.get_master ~__context in
+      let master_pbds = PBDSet.of_list (Db.Host.get_PBDs ~__context ~self:master) in
+      (* Operation run on shared SRs depend on the first pbd of the list to be
+         on the master. *)
+      let sr_master_pbds, rest_pbds = PBDSet.partition (fun pbd ->
+          PBDSet.mem pbd master_pbds) pbds in
+      List.rev_append (PBDSet.elements sr_master_pbds) (PBDSet.elements rest_pbds)
+    else
+      pbds_to_consider
+  in
+  match pbd_candidates with
+  | [] ->
+    let err = if consider_unplugged_pbds then Api_errors.sr_no_pbds else Api_errors.sr_no_plugged_pbds in
```

Comment:
Do we need to change the error reporting here? This is essentially an API change, which has not been justified.

Furthermore, the `SR_NO_PBDs` error has doc string `"The SR has no attached PBDs"` (where "attached" is the same as "plugged"), so the old error actually meant the same as your new error code.

Finally, if we really need a new error, the besides defining it in `api_errors.ml`, you must also add it to `datamodel_errors.ml`.

Reply:
okay, i will remove this extra error since it is not adding much information